internal/codeintel/codenav/internal/lsifstore/locations_by_position.go

```diff
@@ -361,96 +273,88 @@ func uniqueByRange(l shared.Location) [4]int {
-	for _, id := range uploadIDs {
-		if path, ok := skipPaths[id]; ok {
+	for _, id := range opts.UploadIDs {
+		if path, ok := opts.SkipPathsByUploadID[id]; ok {
 			skipConds = append(skipConds, sqlf.Sprintf("(%s, %s)", id, path))
 		}
 	}
 	if len(skipConds) == 0 {
 		skipConds = append(skipConds, sqlf.Sprintf("(%s, %s)", -1, ""))
 	}
 
+	rangesColumn := sqlf.Sprintf(opts.UsageKind.RangesColumnName())
 	query := sqlf.Sprintf(
-		minimalBulkSymbolUsagesQuery,
-		pq.Array(lookupSymbols),
-		pq.Array(uploadIDs),
-		sqlf.Sprintf(usageKind.RangesColumnName()),
-		sqlf.Sprintf(usageKind.RangesColumnName()),
+		symbolUsagesQuery,
+		pq.Array(opts.LookupSymbols),
+		pq.Array(opts.UploadIDs),
+		rangesColumn,
+		rangesColumn,
 		sqlf.Join(skipConds, ", "),
 	)
 
-	locationData, err := s.scanDeduplicatedUploadLoci(s.db.Query(ctx, query))
+	usageData, err := s.scanUploadSymbolLoci(s.db.Query(ctx, query))
 	if err != nil {
 		return nil, 0, err
 	}
 
 	totalCount = 0
-	for _, data := range locationData {
+	for _, data := range usageData {
 		totalCount += len(data.Loci)
 	}
 	trace.AddEvent("TODO Domain Owner",
-		attribute.Int("numUploads", len(locationData)),
+		attribute.Int("numUniqueUploadIDSymbolPairs", len(usageData)),
 		attribute.Int("totalCount", totalCount))
 
-	locations := make([]shared.Location, 0, min(totalCount, limit))
+	usages := make([]shared.Usage, 0, min(totalCount, opts.Limit))
+	offset := opts.Offset
 outer:
-	for _, uploadLoci := range locationData {
-		for _, locus := range uploadLoci.Loci {
+	for _, uploadSymbolLoci := range usageData {
+		for _, locus := range uploadSymbolLoci.Loci {
 			offset--
 			if offset >= 0 {
 				continue
 			}
 
-			locations = append(locations, shared.Location{
-				UploadID: uploadLoci.UploadID,
+			usages = append(usages, shared.Usage{
+				UploadID: uploadSymbolLoci.UploadID,
 				Path:     locus.Path,
 				Range:    shared.TranslateRange(locus.Range),
+				Symbol:   uploadSymbolLoci.Symbol,
+				Kind:     opts.UsageKind,
 			})
 
-			if len(locations) >= limit {
+			if len(usages) >= opts.Limit {
 				break outer
 			}
 		}
 	}
-	trace.AddEvent("TODO Domain Owner", attribute.Int("numLocations", len(locations)))
+	trace.AddEvent("TODO Domain Owner", attribute.Int("numUsages", len(usages)))
 
-	return locations, totalCount, nil
+	return usages, totalCount, nil
 }
 
-const minimalBulkSymbolUsagesQuery = `
+// symbolUsagesQuery gets ALL usages of a bunch of symbols across the ENTIRE instance
+// (within the given set of uploadIDs). We need to do this because the ranges are
+// stored using a custom binary encoding which means we can't use LIMIT+OFFSET at
+// the level of locations.
+const symbolUsagesQuery = `
 WITH RECURSIVE
 ` + symbolIDsCTEs + `
 SELECT
 	ss.upload_id,
-	%s,
-	document_path
+	msn.symbol_name,
+	array_agg(%s ORDER BY document_path),
+	array_agg(document_path ORDER BY document_path)
+    -- ORDER BY ss.upload_id, msn.symbol_name, document_path to maintain determinism for pagination
```

Comment:
```suggestion
	array_agg(%s ORDER BY dl.document_path),
	array_agg(document_path ORDER BY dl.document_path)
    -- ORDER BY ss.upload_id, msn.symbol_name, dl.document_path to maintain determinism for pagination
```

We've been qualifiying all the other column names in here